lib/web/sessions.go

```diff
@@ -827,22 +845,51 @@ func (s *sessionCache) watchWebSessionsOnce(ctx context.Context, reset func()) e
 			if !ok {
 				s.log.
 					WithField("resource_type", fmt.Sprintf("%T", event.Resource)).
-					Warn("sessionCache WebSession watcher received unexpected resource type")
+					Warn("sessionCache: Received unexpected resource type")
+				continue
+			}
+			if !session.GetHasDeviceExtensions() {
+				s.log.
+					WithField("session_id", session.GetName()).
+					Debug("sessionCache: Updated session doesn't have device extensions, skipping")
+				notifyProcessed()
 				continue
 			}
 
-			// If the session got an update, assume it got augmented with device
-			// extensions and clear the cache.
-			// We could be more selective here, but this is good enough.
-			if err := s.releaseResources(session.GetUser(), session.GetName()); err != nil {
+			// Release existing and non-device-aware session.
+			if err := s.releaseResourcesIfNoDeviceExtensions(session.GetUser(), session.GetName()); err != nil {
 				s.log.
 					WithError(err).
-					Debug("sessionCache WebSession wacher: failed to release updated session.")
+					WithField("session_id", session.GetName()).
+					Debug("sessionCache: Failed to release updated session")
 			}
+
+			notifyProcessed()
 		}
 	}
 }
 
+func (s *sessionCache) releaseResourcesIfNoDeviceExtensions(user, sessionID string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	id := sessionKey(user, sessionID)
+	switch sessionCtx, ok := s.sessions[id]; {
+	case !ok:
+		return nil // Session not found
+	case sessionCtx.cfg.Session.GetHasDeviceExtensions():
+		s.log.
+			WithField("session_id", sessionID).
+			Debug("sessionCache: Session already has device extensions, skipping")
+		return nil
+	}
+
+	s.log.
+		WithField("session_id", sessionID).
+		Debug("sessionCache: Releasing session resources due to device extensions upgrade")
+	return s.releaseResourcesLocked(user, sessionID)
```

Comment:
I believe this could use the gentler `s.removeSessionContextLocked(user, sessionID)` (as could the [webconfirm handler][1]). Technically this only ever happens once, shortly before login, so I think it should be fine. Let me know your thoughts on it.

[1]: https://github.com/gravitational/teleport/blob/453bb211d4fa07de7302d33572d82d8604ac816e/lib/web/device_trust.go#L74